File: src/main/java/brs/services/impl/TransactionServiceImpl.java
package brs.services.impl;

import brs.*;
import brs.services.AccountService;
import brs.services.TransactionService;

public class TransactionServiceImpl implements TransactionService {
  private final DependencyProvider dp;

  public TransactionServiceImpl(DependencyProvider dp) {
    this.dp = dp;
  }

  @Override
  public boolean verifyPublicKey(Transaction transaction) {
    Account account = dp.accountService.getAccount(transaction.getSenderId());
    if (account == null) {
      return false;
    }
    if (transaction.getSignature() == null) {
      return false;
    }
    return account.setOrVerify(dp, transaction.getSenderPublicKey(), transaction.getHeight());
  }

  @Override
  public void validate(Transaction transaction) throws BurstException.ValidationException {
    for (Appendix.AbstractAppendix appendage : transaction.getAppendages()) {
      appendage.validate(transaction);
    }
    long minimumFeeNQT = transaction.getType().minimumFeeNQT(dp.blockchain.getHeight(), transaction.getAppendagesSize());
    if (transaction.getFeeNQT() < minimumFeeNQT) {
      throw new BurstException.NotCurrentlyValidException(String.format("Transaction fee %d less than minimum fee %d at height %d",
          transaction.getFeeNQT(), minimumFeeNQT, dp.blockchain.getHeight()));
    }
  }

  @Override
  public boolean applyUnconfirmed(Transaction transaction) {
    Account senderAccount = dp.accountService.getAccount(transaction.getSenderId());
    return senderAccount != null && transaction.getType().applyUnconfirmed(transaction, senderAccount);
  }

  @Override
  public void apply(Transaction transaction) {
    Account senderAccount = dp.accountService.getAccount(transaction.getSenderId());
    senderAccount.apply(dp, transaction.getSenderPublicKey(), transaction.getHeight());
    Account recipientAccount = dp.accountService.getOrAddAccount(transaction.getRecipientId());
    for (Appendix.AbstractAppendix appendage : transaction.getAppendages()) {
      appendage.apply(transaction, senderAccount, recipientAccount);
    }
  }

  @Override
  public void undoUnconfirmed(Transaction transaction) {
    final Account senderAccount = dp.accountService.getAccount(transaction.getSenderId());
    transaction.getType().undoUnconfirmed(transaction, senderAccount);
  }

}

File: src/main/kotlin/brs/services/impl/TransactionServiceImpl.kt
package brs.services.impl

import brs.entity.DependencyProvider
import brs.entity.Transaction
import brs.services.TransactionService
import brs.util.BurstException
import brs.util.convert.toUnsignedString
import brs.util.logging.safeError
import org.slf4j.Logger
import org.slf4j.LoggerFactory

class TransactionServiceImpl(private val dp: DependencyProvider) : TransactionService {
    override fun verifyPublicKey(transaction: Transaction): Boolean {
        val account = dp.accountService.getAccount(transaction.senderId)
        if (account == null) {
            logger.safeError { "Account was null for tx. Account: ${transaction.senderId.toUnsignedString()}" }
            return false
        }
        return transaction.signature != null && dp.db.accountStore.setOrVerify(account, transaction.senderPublicKey, transaction.height)
    }

    override fun validate(transaction: Transaction, preValidate: Boolean) {
        for (appendage in transaction.appendages) {
            appendage.validate(transaction)
        }
        if (preValidate) preValidate(transaction, dp.blockchainService.height)
    }

    override fun preValidate(transaction: Transaction, height: Int) {
        val minimumFeePlanck = transaction.type.minimumFeePlanck(height, transaction.appendagesSize)
        if (transaction.feePlanck < minimumFeePlanck) {
            throw BurstException.NotCurrentlyValidException("Transaction fee ${transaction.feePlanck} less than minimum fee $minimumFeePlanck at height ${dp.blockchainService.height}")
        }
        for (appendage in transaction.appendages) {
            appendage.preValidate(transaction, height)
        }
    }

    override fun applyUnconfirmed(transaction: Transaction): Boolean {
        val senderAccount = dp.accountService.getAccount(transaction.senderId) ?: return false
        return transaction.type.applyUnconfirmed(transaction, senderAccount)
    }

    override fun apply(transaction: Transaction) {
        val senderAccount = dp.accountService.getAccount(transaction.senderId)!!
        senderAccount.apply(dp, transaction.senderPublicKey, transaction.height)
        val recipientAccount = dp.accountService.getOrAddAccount(transaction.recipientId)
        for (appendage in transaction.appendages) {
            appendage.apply(transaction, senderAccount, recipientAccount)
        }
    }

    override fun undoUnconfirmed(transaction: Transaction) {
        val senderAccount = dp.accountService.getAccount(transaction.senderId)!!
        transaction.type.undoUnconfirmed(transaction, senderAccount)
    }

    companion object {
        private val logger: Logger = LoggerFactory.getLogger(TransactionServiceImpl::class.java)
    }
}
